File: app/src/main/java/com/krystianwsul/checkme/gui/tasks/MonthlyDayScheduleEntry.java
package com.krystianwsul.checkme.gui.tasks;

import android.content.Context;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.loaders.CreateTaskLoader;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.ScheduleType;
import com.krystianwsul.checkme.utils.Utils;
import com.krystianwsul.checkme.utils.time.Date;
import com.krystianwsul.checkme.utils.time.TimePair;
import com.krystianwsul.checkme.utils.time.TimePairPersist;

import junit.framework.Assert;

import java.util.Collections;
import java.util.Map;

class MonthlyDayScheduleEntry extends ScheduleEntry {
    private final int mMonthDayNumber;
    private final boolean mBeginningOfMonth;

    @NonNull
    private final TimePair mTimePair;

    MonthlyDayScheduleEntry(@NonNull CreateTaskLoader.ScheduleData.MonthlyDayScheduleData monthlyDayScheduleData) {
        mMonthDayNumber = monthlyDayScheduleData.getDayOfMonth();
        mBeginningOfMonth = monthlyDayScheduleData.getBeginningOfMonth();
        mTimePair = monthlyDayScheduleData.getTimePair().copy();
    }

    private MonthlyDayScheduleEntry(int monthDayNumber, boolean beginningOfMonth, @NonNull TimePair timePair, @Nullable String error) {
        super(error);

        mMonthDayNumber = monthDayNumber;
        mBeginningOfMonth = beginningOfMonth;
        mTimePair = timePair;
    }

    MonthlyDayScheduleEntry(@NonNull ScheduleDialogFragment.ScheduleDialogData scheduleDialogData) {
        Assert.assertTrue(scheduleDialogData.getMScheduleType() == ScheduleType.MONTHLY_DAY);
        Assert.assertTrue(scheduleDialogData.getMMonthlyDay());

        mMonthDayNumber = scheduleDialogData.getMMonthDayNumber();
        mBeginningOfMonth = scheduleDialogData.getMBeginningOfMonth();
        mTimePair = scheduleDialogData.getMTimePairPersist().getTimePair();
    }

    @NonNull
    @Override
    public String getText(@NonNull Map<CustomTimeKey, CreateTaskLoader.CustomTimeData> customTimeDatas, @NonNull Context context) {
        String day = Utils.ordinal(mMonthDayNumber) + " " + context.getString(R.string.monthDay) + " " + context.getString(R.string.monthDayStart) + " " + context.getResources().getStringArray(R.array.month)[mBeginningOfMonth ? 0 : 1] + " " + context.getString(R.string.monthDayEnd);

        if (mTimePair.mCustomTimeKey != null) {
            Assert.assertTrue(mTimePair.mHourMinute == null);

            CreateTaskLoader.CustomTimeData customTimeData = customTimeDatas.get(mTimePair.mCustomTimeKey);
            Assert.assertTrue(customTimeData != null);

            return day + ", " + customTimeData.getName();
        } else {
            Assert.assertTrue(mTimePair.mHourMinute != null);

            return day + ", " + mTimePair.mHourMinute.toString();
        }
    }

    @NonNull
    @Override
    public CreateTaskLoader.ScheduleData getScheduleData() {
        return new CreateTaskLoader.ScheduleData.MonthlyDayScheduleData(mMonthDayNumber, mBeginningOfMonth, mTimePair);
    }

    @NonNull
    @Override
    public ScheduleDialogFragment.ScheduleDialogData getScheduleDialogData(@NonNull Date today, @Nullable CreateTaskActivity.ScheduleHint scheduleHint) {
        Date date = (scheduleHint != null ? scheduleHint.getMDate() : today);

        date = Utils.getDateInMonth(date.getYear(), date.getMonth(), mMonthDayNumber, mBeginningOfMonth);

        return new ScheduleDialogFragment.ScheduleDialogData(date, Collections.singleton(date.getDayOfWeek()), true, mMonthDayNumber, (mMonthDayNumber - 1) / 7 + 1, date.getDayOfWeek(), mBeginningOfMonth, new TimePairPersist(mTimePair), ScheduleType.MONTHLY_DAY);
    }

    @NonNull
    @Override
    public ScheduleType getScheduleType() {
        return ScheduleType.MONTHLY_DAY;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeInt(mMonthDayNumber);
        parcel.writeInt(mBeginningOfMonth ? 1 : 0);
        parcel.writeParcelable(mTimePair, 0);
        parcel.writeString(getError());
    }

    @SuppressWarnings("unused")
    public static final Creator<MonthlyDayScheduleEntry> CREATOR = new Creator<MonthlyDayScheduleEntry>() {
        @Override
        public MonthlyDayScheduleEntry createFromParcel(Parcel in) {
            int monthDayNumber = in.readInt();

            boolean beginningOfMonth = (in.readInt() == 1);

            TimePair timePair = in.readParcelable(TimePair.class.getClassLoader());
            Assert.assertTrue(timePair != null);

            String error = in.readString();

            return new MonthlyDayScheduleEntry(monthDayNumber, beginningOfMonth, timePair, error);
        }

        @Override
        public MonthlyDayScheduleEntry[] newArray(int size) {
            return new MonthlyDayScheduleEntry[size];
        }
    };
}

File: app/src/main/java/com/krystianwsul/checkme/gui/tasks/MonthlyDayScheduleEntry.kt
package com.krystianwsul.checkme.gui.tasks

import android.content.Context
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.utils.CustomTimeKey
import com.krystianwsul.checkme.utils.ScheduleType
import com.krystianwsul.checkme.utils.Utils
import com.krystianwsul.checkme.utils.time.Date
import com.krystianwsul.checkme.utils.time.TimePairPersist
import com.krystianwsul.checkme.viewmodels.CreateTaskViewModel


class MonthlyDayScheduleEntry(monthlyDay: CreateTaskViewModel.ScheduleData.MonthlyDay) : ScheduleEntry() {

    private val monthDayNumber = monthlyDay.dayOfMonth
    private val beginningOfMonth = monthlyDay.beginningOfMonth
    private val timePair = monthlyDay.timePair

    override val scheduleData: CreateTaskViewModel.ScheduleData
        get() = CreateTaskViewModel.ScheduleData.MonthlyDay(monthDayNumber, beginningOfMonth, timePair)

    override val scheduleType = ScheduleType.MONTHLY_DAY

    override fun getText(customTimeDatas: Map<CustomTimeKey<*>, CreateTaskViewModel.CustomTimeData>, context: Context): String {
        val day = Utils.ordinal(monthDayNumber) + " " + context.getString(R.string.monthDay) + " " + context.getString(R.string.monthDayStart) + " " + context.resources.getStringArray(R.array.month)[if (beginningOfMonth) 0 else 1] + " " + context.getString(R.string.monthDayEnd)

        return "$day, " + if (timePair.customTimeKey != null) {
            check(timePair.hourMinute == null)

            val customTimeData = customTimeDatas.getValue(timePair.customTimeKey)

            customTimeData.name
        } else {
            timePair.hourMinute!!.toString()
        }
    }

    override fun getScheduleDialogData(today: Date, scheduleHint: CreateTaskActivity.Hint.Schedule?): ScheduleDialogFragment.ScheduleDialogData {
        var date = scheduleHint?.date ?: today

        date = Utils.getDateInMonth(date.year, date.month, monthDayNumber, beginningOfMonth)

        return ScheduleDialogFragment.ScheduleDialogData(date, mutableSetOf(date.dayOfWeek), true, monthDayNumber, (monthDayNumber - 1) / 7 + 1, date.dayOfWeek, beginningOfMonth, TimePairPersist(timePair), ScheduleType.MONTHLY_DAY)
    }
}
